Knowledge distillation tutorial initial commit (#25182)
huggingface/transformers: docs/source/en/tasks/knowledge_distillation_for_image_classification.md

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->
# Knowledge Distillation for Computer Vision

> Knowledge distillation is a technique used to transfer knowledge from a larger, more complex model (teacher) to a smaller, simpler model (student). In the context of image classification, the goal is to train the student model to mimic the behavior of the teacher model. In this notebook, we will do task-specific knowledge distillation. We will use [beans](https://huggingface.co/datasets/beans) for this.

Reviewer: Worth linking the OG Knowledge Distillation paper: https://arxiv.org/abs/1503.02531

> Knowledge distillation is a technique used to transfer knowledge from a larger, more complex model (teacher) to a smaller, simpler model (student). In the context of image classification, the goal is to train the student model to mimic the behavior of the teacher model. In this notebook, we will do task-specific knowledge distillation. We will use [beans](https://huggingface.co/datasets/beans) for this.

Reviewer 2: Would be nice to also add the special `[[open-in-colab]]` at the top so a notebook is automatically generated :)

> This tutorial aims to demonstrate how to distil ResNet (teacher model)[https://huggingface.co/microsoft/resnet-50] to [MobileNet](https://huggingface.co/google/mobilenet_v2_1.4_224) (student model) using `Trainer` API of 🤗 Transformers.

Reviewer: ```suggestion

This tutorial aims to demonstrate how to distill a (https://huggingface.co/papers/1512.03385) (teacher model) to a [MobileNetV2](https://huggingface.co/papers/1801.04381) (student model) using `Trainer` API of 🤗 Transformers.
```

Let's install the libraries needed for distillation and evaluating the process. 

```bash
pip install transformers datasets accelerate tensorboard evaluate --upgrade
```

> In this example, we are using `google/mobilenet_v2_1.4_224` and `microsoft/resnet-50`, both are trained with ImageNet-1k dataset on resolution of 224x224. You can see that feature extractor returns the same output below.

Reviewer: The student model is not supposed to be initialized from pre-trained checkpoints, though.

```python
from transformers import AutoFeatureExtractor
from PIL import Image
import requests
import numpy as np

teacher_extractor = AutoFeatureExtractor.from_pretrained("microsoft/resnet-50")
student_extractor = AutoFeatureExtractor.from_pretrained("google/mobilenet_v2_1.4_224")

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
sample = Image.open(requests.get(url, stream=True).raw)

np.array_equal(teacher_extractor(sample),student_extractor(sample))
```

We will now load and pre-process the dataset. 

```python
from datasets import load_dataset

dataset = load_dataset("beans")
```

> We can use either of the processors given they return the same output. We will use `map()` method of `dataset` to apply the preprocessing to every split of the dataset.

Reviewer: > We can use either of the processors given they return the same output.

Do they use the exact same normalization of pixels? They actually don't. Looking at the preprocessor configs ([here](https://huggingface.co/microsoft/resnet-50/blob/main/preprocessor_config.json) and [here](https://huggingface.co/google/mobilenet_v2_1.4_224/blob/main/preprocessor_config.json)), we can see that their `image_mean` and `image_std` stats are different. So, we need to be careful here. 

If we're randomly initializing the student model (which I think we should for distillation), then we need to either calculate the `image_mean` and `image_std` stats from the underlying dataset (`beans` in this case). 

Or, we could just scale the pixel values to be either in the [0, 1] or [-1, 1] range. 

Also, the teacher model needs to be fine-tuned on the `beans` dataset here. We shouldn't be using an ImageNet-1k pre-trained model here.

```python
def process(examples):
    processed_inputs = teacher_extractor(examples["image"])
    return processed_inputs

processed_datasets = dataset.map(process, batched=True)
```

> Essentially, we want the student to mimick the teacher so we will get the output from the teacher and the student, multiply it with a parameter called `temperature`, then calculate KL-divergence between them. We will then calculate loss based on KL-divergence and student output, and to provide a trade-off between them, we will use a parameter called `lambda`. In this tutorial, we are going to use `lambda` as 0.5 and temperature as 5.

Reviewer 2: ```suggestion
To get the student to mimic the teacher, multiply the output from both models with a parameter called `temperature` and then calculate the KL-divergence between them. The loss is calculated based on the KL-divergence and student output. To provide a trade-off between the KL-divergence and student output, we will use a parameter called `lambda`. 
```

```python
from transformers import TrainingArguments, Trainer
import torch
import torch.nn as nn
import torch.nn.functional as F

> class ImageDistilTrainer(Trainer):

Reviewer 2: ```suggestion
class ImageDistillTrainer(Trainer):
```

    def __init__(self, *args, teacher_model=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.teacher = teacher_model
        self.student = student_model
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        self.teacher.to(device)
        self.teacher.eval()

    def compute_loss(self, student, inputs, return_outputs=False):
        lambda_param = 0.5

>         temperature = 5

Reviewer: Should be taken from the user. Let's define them as arguments of the trainer class.

        student_output = self.student(**inputs)

        with torch.no_grad():
          teacher_output = self.teacher(**inputs)

        loss_function = nn.KLDivLoss(reduction="batchmean")
        loss_logits = (loss_function(
            F.log_softmax(student_output.logits / temperature, dim=-1),

>             F.softmax(teacher_output.logits / temperature, dim=-1)) * (temperature ** 2))

Reviewer: Why is the first one log_softmax()?

        loss = lambda_param * student_output.loss + (1. - lambda_param) * loss_logits
        return (loss, student_output) if return_outputs else loss
```

> We will now login to Hugging Face Hub so we can push out model to Hugging Face Hub through `Trainer`. 

Reviewer 2: ```suggestion
We will now log-in to push our model to the Hub with [`Trainer`]. 
```

```python
from huggingface_hub import notebook_login

notebook_login()
```

> Let's initialize the data collator and `Trainer`. We will use `DefaultDataCollator` for this. 

Reviewer 2: ```suggestion
Let's initialize the [`TrainingArguments`], [`DefaultDataCollator`], and the student and teacher models:
```

```python
from transformers import AutoModelForImageClassification, DefaultDataCollator

training_args = TrainingArguments(
    output_dir="my-awesome-model",
    num_train_epochs=5,
    fp16=True,
    logging_dir=f"{repo_name}/logs",
    logging_strategy="epoch",
    evaluation_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    metric_for_best_model="accuracy",
    report_to="tensorboard",
    push_to_hub=True,
    hub_strategy="every_save",
    hub_model_id=repo_name,
    )

data_collator = DefaultDataCollator()

num_labels = len(processed_datasets["train"].features["labels"].names)

# initialize models
teacher_model = AutoModelForImageClassification.from_pretrained(
    "microsoft/resnet-50",
    num_labels=num_labels,
    ignore_mismatched_sizes=True
)

student_model = AutoModelForImageClassification.from_pretrained(
    "google/mobilenet_v2_1.4_224",
    num_labels=num_labels,
    ignore_mismatched_sizes=True
)
```

> We can use `compute_metrics` function to evaluate our model on the test set. This function will be used during the training process to compute the `accuracy` & `f1` of our model.

Reviewer 2: ```suggestion
We can use the `compute_metrics` function to evaluate our model on the test set. This function is used during training to compute the `accuracy` and `f1` scores.
```

```python
import evaluate
import numpy as np

accuracy = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    acc = accuracy.compute(references=labels, predictions=np.argmax(predictions, axis=1))
    return {"accuracy": acc["accuracy"]}
```

> Let's initialize the `Trainer` with the training arguments we defined. 

Reviewer 2: ```suggestion
Initialize the `ImageDistillTrainer` with the training arguments we defined:
```

```python
trainer = ImageDistilTrainer(
    student_model,
    training_args,
    teacher_model=teacher_model,
    train_dataset=processed_datasets["train"],
    eval_dataset=processed_datasets["validation"],
    data_collator=data_collator,
    tokenizer=teacher_extractor,
    compute_metrics=compute_metrics,
)
```

We can now train our model.

```python
trainer.train()
```

> We can evaluate the model on the test set.

Reviewer 2: ```suggestion
To evaluate the model on the test set:
```

```python
trainer.evaluate(processed_datasets["test"])
```

> On test set, our model reaches 88 percent evaluation accuracy. To have a sanity check over efficiency of distillation, we also trained MobileNet on beans dataset from scratch with same hyperparameters and observed 33 percent accuracy on test set.

Reviewer 2: ```suggestion
On the test set, the model reaches 88% evaluation accuracy. To compare distillation efficiency, we also trained MobileNet on the beans dataset from scratch with the same hyperparameters and observed 33% accuracy on the test set.
```
